### ISSUE TITLE
[Bug]: Unable to forward remotehost:port 

### ISSUE THREAD

**Issue by sibidharan** [coder/code-server#5723]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### OS/Web Information

- Web Browser: Safari
- Local OS: MacOS
- Remote OS: Linux
- Remote Architecture: amd64
- `code-server --version`: 4.8.0

### Steps to Reproduce

1. Spawn a service in a neighbour container
2. Under ports, paste container-ip:host
3. It tells remotehost:port is forwarded, but it is unable to do

When I open the forwarded URL proxy, it says 

```
connect ECONNREFUSED 0.0.0.0:8080
```

### Expected

The service running at remotehost:port should be forwarded via web proxy 

### Actual

Only locally running ports are being forwarded and remorehost:port is not being forwarded unlike VS Code Desktop. It is not even respecting the host part and just attempting to forward a locally running port.

### Does this issue happen in VS Code or GitHub Codespaces?

- [X] I cannot reproduce this in VS Code.
- [X] I cannot reproduce this in GitHub Codespaces.

### Are you accessing code-server over HTTPS?

- [X] I am using HTTPS.

**Comment by jsjoeio** [comment]
Can you elaborate "Spawn a service in a neighbour container"? Then I can try and reproduce this locally.

**Comment by aureliainnovatives** [comment]
I am facing relatively same problem. If I run a Node App on port 3000 via Code-Server container and NGinx Proxy module, it's not accessible via url/proxy/3000 port. How to get rid of proxy/ and allow it like URL:3000 ?

**Comment by jsjoeio** [comment]
Hmm 🤔 @code-asher any ideas on how this should work behind nginx?

**Comment by sibidharan** [comment]
> Can you elaborate "Spawn a service in a neighbour container"? Then I can try and reproduce this locally.

Hi,

For example, in a docker compose, I spawn 3 containers, one is MySQL, another one is Adminer and another one is my Apache+PHP+Codeserver containers. In this 3 containers are spawned, and they are reachable within themselves. If I forward container-ip:3306 in my desktop vs code over SSH, it forwards container-ip:3306 over SSH to my localhost:3306. 

If my codeserver can see any host:IP forwarded, it should be able to. For now, it is just looking in localhost and not for the servers running on the same network, even if I specify host/ip. But this works correctly in my Desktop VS Code.

In the below example, a MongoDB service is running on that IP:port and my desktop VS code is able to port forward that. 

<img width="1136" alt="Screenshot 2022-11-01 at 5 28 17 AM" src="https://user-images.githubusercontent.com/10979210/199131108-8be1aa4f-7321-4a0e-93e5-e28784f30836.png">

If not MongoDB, the code-server should be able to forward HTTP, like adminer.

**Comment by jsjoeio** [comment]
Thanks for elaborating! I'm not sure how'd we implement this on the code-server side 🤔

@code-asher any ideas?

**Comment by code-asher** [comment]
I believe we have the proxy hardcoded to localhost right now so we 
will need to fix that.

**Comment by code-asher** [comment]
Uh I might have misread the question.  For implementing maybe we can leave the IP alone if it is 172.*?

**Comment by code-asher** [comment]
Actually we should just always leave the IP alone I think and when it is missing then we default to localhost.

**Comment by code-asher** [comment]
But that means we need to encode the IP into the proxy URL somehow.  `/proxy/3000?ip=[ip]` or `/proxy/[ip]/3000` for example.

**Comment by sibidharan** [comment]
> Uh I might have misread the question. For implementing maybe we can leave the IP alone if it is 172.*?

I don't think so.. If the IP is reachable, code-server should forward it. I don't know how protocols like MySQL or MongoDB work in HTTPS.. There should be some way!

**Comment by benz0li** [comment]
> But that means we need to encode the IP into the proxy URL somehow. `/proxy/3000?ip=[ip]` or `/proxy/[ip]/3000` for example.

@code-asher Does this affect https://github.com/coder/code-server/blob/main/patches/proxy-uri.diff?

**Comment by code-asher** [comment]
Ah yeah good point, I think we would need to add something like `{{ip}}` to `VSCODE_PROXY_URI`?  But to prevent breakage maybe we can default to localhost when the endpoint is `/proxy/{{ip}}/3000` or whatever format we settle on.

**Comment by benz0li** [comment]
> Ah yeah good point, I think we would need to add something like `{{ip}}` to `VSCODE_PROXY_URI`?

Only if format `/proxy/{{ip}}/{{port}}/[path/to/page]` is settled on. 

> But to prevent breakage maybe we can default to localhost when the endpoint is `/proxy/{{ip}}/3000` or whatever format we settle on.

If format `/proxy/{{port}}/[path/to/page]?ip=192.0.2.2`[^1] is settled on and query string parameter `ip` is unset it defaults to `localhost`.

[^1]: https://datatracker.ietf.org/doc/html/rfc5737

**Comment by benz0li** [comment]
@code-asher @jsjoeio Please make sure that proxying is restricted to the following ranges:

**IPv4**

* `10.0.0.0/8`[^1]
* `172.16.0.0/12`[^1]
* `192.168.0.0/16`[^1]

plus `127.0.0.1`[^2]/`localhost`.

**IPv6**

* `fc00::/7`[^3]

plus `::1`[^4]/`localhost`.

[^1]: https://en.wikipedia.org/wiki/IPv4#Private_networks
[^2]: https://en.wikipedia.org/wiki/IPv4#Loopback
[^3]: https://en.wikipedia.org/wiki/IPv6_address#Unique_local_addresses
[^4]: https://en.wikipedia.org/wiki/IPv6_address#Local_addresses

Proxying to arbitrary IPs must be prevented.  
👉 Otherwise, this opens up an incredible number of (undesirable) use cases.

**Comment by benz0li** [comment]
I prefer format `/proxy/{{port}}/[path/to/page]?ip=192.0.2.2`.

**Comment by code-asher** [comment]
Great points and thank you for the links!! 

I prefer the query parameter as well but I am concerned that the 
application itself might use query parameters and have one called 
`ip` or maybe it is an app with multiple pages and navigating to a 
page drops the `ip` query parameter.

If we embed it into the path on the other hand the application 
would not be able to mess it up.

We could make it a requirement that applications running behind 
code-server's proxy specially treat the `ip` query variable if it 
exists but I could see that not going too well. XD

**Comment by benz0li** [comment]
> I prefer the query parameter as well but I am concerned that the application itself might use query parameters and have one called `ip` or maybe it is an app with multiple pages and navigating to a page drops the `ip` query parameter. 

Understood.

> If we embed it into the path on the other hand the application would not be able to mess it up.

True.

> We could make it a requirement that applications running behind code-server's proxy specially treat the `ip` query variable if it exists but I could see that not going too well. XD

Agreed.

**Comment by benz0li** [comment]
@code-asher I do not like to modify the `VSCODE_PROXY_URI` (again). Applications and/or extension making use of `VSCODE_PROXY_URI` are running on `127.0.0.1`/`localhost` by default.

It should be possible to differentiate between `/proxy/{{port}}/[path/to/page]` and `/proxy/{{ip}}/{{port}}/[path/to/page]` and proxy the former to `127.0.0.1`/`localhost` and the latter to `ip`.  
ℹ️  IPv4: If the part after the second `/` does not contain a `.`, it is already the `port`– no `ip` set; thus forward to `127.0.0.1`/`localhost`.

**Comment by benz0li** [comment]
> It says the server's public IP (95.111.198.224) of my 172.0.0.5 (that's on wireguard), because VS Code Desktop is able to forward any arbitrary IP) - and this is like I am opening http://whatismyip.selfmade.ninja from 95.111.198.224

Since VS Code Desktop [and the computer it runs on] is under your control, that is fine.

> This opens up a lot of opportunities for developers, and I don't understand what undesirable things will happen by forwarding arbitrary IPs.

Because if code-server is deployed as SaaS, users would be able to use your server as a proxy for pretty much anything.  
ℹ️ For this reason this functionality should be restricted to private subnets.

**Comment by sibidharan** [comment]
> Because if code-server is deployed as SaaS, users would be able to use your server as a proxy for pretty much anything.
> ℹ️ For this reason this functionality should be restricted to private subnets.

This is a potential security risk, agreed. 

But a user can do this whatsoever if he has a shell, by doing an SSH port forward. So why don't we let the admins decide if the code-server can allow forward on all IP ranges or only private IP with a configuration like `allowed-ips` as an array that takes CIDR - and it can default to Class A,B&C, so unless modified, it won't allow forwarding.?

**Comment by code-asher** [comment]
I like the idea of a whitelist that you can extend or override 
through the config.yaml or command-line flags although it does 
seem pretty easy to bypass.  You could spin up or write your own 
proxy, access it through code-server's proxy, then access whatever 
you want anyway.

@benz0li I like how you think.  So we will leave 
`VSCODE_PROXY_URI` alone.  If for some reason 
applications/extensions do start needing to specify a 
non-localhost IP we can revisit.  The algorithm you laid out for 
differentiating looks great to me as well.

@sibidharan Ahh I skipped over the part about non-HTTP protocols, 
thanks for pointing that out!  Your solution makes sense to me.

**Comment by Toskies** [comment]
The service I deployed is in Kubernetes' pod, and the localaddress provided by the codeserver auto forward port is https://<domain name>/notebook/northstar-team/test-isaacsim/proxy/8211/ 
When I use this link to access, I reported an error not found.
My normal link to access the service is  http://<ip>:8211/streaming/client
I tried https://<domain name>/notebook/northstar-team/test-isaacsim/proxy/8211/streaming/client , But it's not the interface I want
![codeserver](https://github.com/coder/code-server/assets/56681146/832d9a31-bdd1-4036-8462-8a6a6278b152)

**Comment by code-asher** [comment]
@Toskies Do you have the same problem if you use `/absproxy/8211/` instead?

You might need to log the requests in your app to see if it is getting what it expects.  Unless the 404 comes from code-server and not your app?

**Comment by sibidharan** [comment]
> You could spin up or write your own proxy, access it through code-server's proxy, then access whatever you want anyway.

Yes, I am able to do this using `socat` and its forwarding anything I want. For example, 

```
socat TCP-LISTEN:8081,reuseaddr,fork TCP:adminer.selfmade.ninja:8080
```

this command could forward adminer.selfmade.ninja:8080 to my localhost's 8081 and I am able to view it via my --proxy-domain like  https://8081.proxy.domain 

I have setup a container with dynamic proxy, and its working like a charm. So adding this as a feature is much appreciated with whitelist feature.

**Comment by sibidharan** [comment]
Any update on this ?

**Comment by code-asher** [comment]
I do not believe anyone is working on this.
